README.md
# Spambayes3 - Python 3 version of spambayes

An attempt to port the sadly neglected spambayes spam analysis
software to Python 3.  The migration was based on spambasyes-1.1a6, the
latest version available (as far as I can tell).

## Progress

scripts/sb_filter.py, scripts/sb_mboxtrain.py and the supporting
modules have been convertered to Python 3 and lightly tested.

### Update to 1.1b3 [2022-06-04]

Merged in updates from a later spambayes release (1.1b3). This later
version was located by ajschorr.

No additional code has been converted to Python 3.

# Original README.TXT

Copyright (C) 2002-2007 Python Software Foundation; All Rights
Reserved

The Python Software Foundation (PSF) holds copyright on all material
in this project.  You may use it under the terms of the PSF license;
see LICENSE.txt.


Overview
========

SpamBayes is a tool used to segregate unwanted mail (spam) from the mail you
want (ham).  Before SpamBayes can be your spam filter of choice you need to
train it on representative samples of email you receive.  After it's been
trained, you use SpamBayes to classify new mail according to its spamminess
and hamminess qualities.

When SpamBayes filters your email, it compares each unclassified message
against the information it saved from training and makes a decision about
whether it thinks the message qualifies as ham or spam, or if it's unsure
about how to classify the message.  It then passes this information on to
your mail client.  Unless you are using IMAP or Outlook, this means it adds
a header to each message, X-SpamBayes-Classification: spam|ham|unsure.  You
can then filter on this header, to file away suspected spam into its own
mail folder for example.  IMAP and Outlook both have the capacity to do the
filtering themselves, so the header is not necessary.

If you have any questions that this document does not answer, you should
definitely try the SpamBayes website <http://spambayes.org>, and in
particular, try reading the list of frequently asked questions:
<http://spambayes.org/faq.html>


Prerequisites
=============

You need to have Python 2.2 or later (2.3 is recommended).  You can
download Python from <http://www.python.org/download/>.
Many distributions of UNIX now ship with Python - try typing 'python'
at a shell prompt.

You also need version 2.4.3 or above of the Python "email" package.
If you're running Python 2.2.3 or above then you already have a good
version of the email package.

If not, you can download email version 2.5 from the email SIG at
<http://www.python.org/sigs/email-sig> and install it - unpack the
archive, cd to the email-2.5 directory and type "python setup.py
install".  This will install it into your Python site-packages
directory.  You'll also need to move aside the standard "email"
library - go to your Python "Lib" directory and rename "email" to
"email_old".

To run the Outlook plug-in from source, you also need have the win32com
extensions installed (win32all-149 or above), which you can get from
<http://starship.python.net/crew/mhammond>.

When installing SpamBayes on some *nix systems, such as Debian, you may need
to install the python-dev package.  This can be done with a command like
"apt-get install python-dev" (this may vary between distributions).


Getting the software
====================

If you don't already have it, you can download the latest release of
SpamBayes from <http://spambayes.org/download.html>.


For the Really Impatient
========================

If you get your mail from a POP3 server, then all you should need to do
to get running is change your mail client to send and receive mail from
"localhost", and then run "python setup.py install" and then
"python scripts/sb_server.py -b" in the directory you expanded the
SpamBayes source into.  This will open a web browser window - click the
"Configuration" link at the top right and fill in the various settings.


Installation
============

The first thing you need to do is run "python setup.py install" in the
directory that you expanded the SpamBayes archive into (to do this, you
probably need to open up a console window/command prompt/DOS prompt,
and navigate to the appropriate directory with the "cd" command).  This
will install all the files that you need into the correct locations.
After this, you can delete that directory; it is no longer required.

Before you begin
----------------

It's a good idea to train SpamBayes before you start using it, although
this isn't compulsory.  You need to save your incoming email for awhile,
segregating it into two piles, known spam (bad mail) and known ham (good
mail).  It's best to train on recent email, because your interests and the
nature of what spam looks like change over time.  Once you've collected a
fair portion of each (anything is better than nothing, but it helps to have
a couple hundred of each), you can tell SpamBayes, "Here's my ham and my
spam".  It will then process that mail and save information about different
patterns which appear in ham and spam.  That information is then used
during the filtering stage.  See the "Training" section below for details.

For more detailed instructions, please read the appropriate section below
(if you don't know, you probably want the POP3 Proxy section).

Outlook plug-in
---------------

For information about how to use the Outlook plug-in, please read the
"about.html" file in the Outlook2000 directory.  If you want to run the
Outlook plug-in from source, you should also read the "README.txt" file
in that directory.


POP3 Proxy
----------

You need to configure your email client to talk to the proxies instead of
the real email servers.  Change your equivalent of "pop3.example.com" to
"localhost" (or to the name of the machine you're running the proxy on) in
your email client's setup, and do the same with your equivalent of
"smtp.example.com".

Now launch SpamBayes, either by running "pop3proxy_service.py install"
and then "net start pop3proxy" (for those using Windows 2000, Windows NT
or Windows XP), or the "sb_server.py" script (for everyone else).  Note
that if you want to use the service, you need to also have Mark
Hammond's win32 extensions for Python installed:

<http://starship.python.net/crew/mhammond/win32/Downloads.html>

All you need to do to configure SpamBayes is to open a web page to
<http://localhost:8880>, click on the "Configuration" link at the top
right, and fill in the relevant details.  Everything should be OK with the
defaults, except for the POP3 and SMTP server information at the top, which
is required. For the local ports to proxy on, if you are only proxying one
server, and you are using Windows, then 110 is probably the best port to
try first.  If that doesn't work, try using 8110 (and if you are proxying
multiple ports, continue with 8111, 8112, and so on).  Note that *nix users
may not have permission to bind ports lower than 1025, so should choose
numbers higher than that.

When you check your mail in your mail client now, messages should have an
addition SpamBayes header (you may not be able to see this by default).
You should be able to create a mail folder called "Spam" and set up a
filtering rule that puts emails with an "X-Spambayes-Classification: spam"
header into that folder.

Note that if you set your mail client to delete the mail without
downloading the whole message (like Outlook Express's "delete from server"
rule) that you may not get accurate results - the classification will
be based on the headers only, not the body.  This is not recommended.


IMAP Filter
-----------

To configure SpamBayes, run "sb_imapfilter.py -b", which should open a web
page to <http://localhost:8880>, click on the "Configuration" link at the
top right, and fill in the relevant details.  Everything should be OK with
the defaults, except for the server information at the top.

You now need to let SpamBayes know which IMAP folders it should work with.
Use the "configure folders to filter" and "configure folders to train"
links on the web page <http://localhost:8880> to do this.  The 'filter'
folders are those that will have mail that you want to identify as either
ham (good) or spam (bad) - this will probably be your Inbox.  The 'train'
folders are those that contain examples of ham and spam, to assist SpamBayes
with its classification.  (Folders can be used for both training and
filtering).

You then need to set the IMAP filter up to run periodically.  At the moment,
you'll need to do this from a command (or DOS) prompt.  You should run the
command "python sb_imapfilter.py -c -t -l 5".  The '-c' means that the script
should classify new mail, the '-t' means that the script should train any
mail that you have told it to, and the '-l 5' means that the script should
execute every five minutes (you can change this as required).


XML-RPC Server
--------------

The XML-RPC server (new in 1.1a4) web interface is almost identical the the
POP3 proxy user interface.  Instead of proxying POP3 communications though
it provides an XML-RPC server your (typically non-mail) applications can use
to score content submissions.

To install and configure it:

1. Unpack and install the distribution:

    tar xvfz spambayes-1.1a4.tar.gz
    cd spambayes-1.1a4
    python setup.py install

2. Devote a runtime directory to it:

    SBDIR=/usr/local/spambayes/core_server  # or whatever...
    mkdir -p $SBDIR

3. Create an INI file:

    cd $SBDIR
    cat > bayescustomize.ini <<EOF
[globals]
verbose:False

[Headers]
include_evidence:True
include_score:True

[Tokenizer]
record_header_absence:True
summarize_email_prefixes:True
summarize_email_suffixes:True
mine_received_headers:True
x-pick_apart_urls:True
x-fancy_url_recognition:False
x-lookup_ip:True
lookup_ip_cache:$SBDIR/dnscache.pck
max_image_size:100000
crack_image_cache:$SBDIR/imagecache.pck

crack_images:True
image_size:True
ocr_engine:gocr
[Classifier]
use_bigrams:False

[Categorization]
ham_cutoff:0.2
spam_cutoff:0.85

[Storage]
persistent_storage_file:$SBDIR/hammie.db
persistent_use_database:pickle
messageinfo_storage_file:$SBDIR/messageinfo.fs

[html_ui]
display_score:True
EOF

4. Finally, start it:

    BAYESCUSTOMIZE=$SBDIR/bayescustomize.ini core_server.py -m XMLRPCPlugin

Note that it creates both a web server (defaulting to localhost:8880) and an
XML-RPC server (defaulting to localhost:8001).


Procmail filtering
------------------

Many people on Unix-like systems have procmail available as an optional or
as the default local delivery agent.  Integrating SpamBayes checking with
Procmail is straightforward.

First, create a SpamBayes database, by running "sb_filter.py -n".  If
you have some mail around that you can use to train it, do you (see the
"command line training" section below).  Note that if you don't, all your
mail will start out as 'unsure'.

Now, create a .spambayesrc file.  There are lots of options you could have
in here, but for the moment, just have these:

    [Storage]
    persistent_use_database = True
    persistent_storage_file = ~/.hammiedb

(Replace the latter with the location of the .hammiedb file that
sb_filter.py created in the first step).

Once you've trained SpamBayes on your
collection of know ham and spam, you can use the sb_filter.py script to
classify incoming mail like so:

    :0 fw:hamlock
    | /usr/local/bin/sb_filter.py

The above Procmail recipe tells it to run /usr/local/bin/sb_filter.py.
Since no command line arguments are given, it relies on the options file
specified by the BAYESCUSTOMIZE variable for all parameters.  While
sb_filter.py is running, Procmail uses the lock file hamlock to prevent
multiple invocations from stepping on each others' toes.  (It's not strictly
necessary in this case since no files on-disk are modified, but Procmail
will still complain if you don't specify a lock file.)

The result of running sb_filter.py in filter mode is that Procmail will
use the output from the run as the mail message for further processing
downstream. sb_filter.py inserts an X-SpamBayes-Classification header in
the output message which looks like:

    X-SpamBayes-Classification: ham; 0.00; '*H*': 1.00; '*S*': 0.00; 'python': 0.00;
	'linux,': 0.01; 'desirable': 0.01; 'cvs,': 0.01; 'perl.': 0.02;
	...

You can then use this to segregate your messages into various inboxes, like
so:

    :0
    * ^X-SpamBayes-Classification: spam
    spam

    :0
    * ^X-SpamBayes-Classification: unsure
    unsure

The first recipe catches all messages which sb_filter.py classified as spam.
The second catches all messages about which it was unsure.  The combination
allows you to isolate spam from your good mail and tuck away messages it was
unsure about so you can scan them more closely.


VM and Gnus
-----------

VM and Gnus are mail readers distributed with Emacs and XEmacs.  The
SpamBayes.el file in the contrib directory contains code and
instructions for VM and Gnus integration.


Training
========

POP3 Proxy
----------

You can train the system through the web interface: <http://localhost:8880>.
Follow the "Review messages" link and you'll see a list of the emails that
the system has seen so far.  Check the appropriate boxes and hit Train.
The messages disappear and if you go back to the home page you'll see that
the "Total emails trained" has increased.

Alternatively, when you receive an incorrectly classified message, you can
forward it to the SMTP proxy for training.  If the message should have been
classified as spam, forward or bounce the message to
spambayes_spam@localhost, and if the message should have been classified as
ham, forward it to spambayes_ham@localhost.  You can still review the
training through the web interface, if you wish to do so.

Note that some mail clients (particularly Outlook Express) do not forward
all headers when you bounce, forward or redirect mail.  For these clients,
you will need to use the web interface to train.

Once you've done this on a few spams and a few hams, you'll find that the
X-Spambayes-Classification header is getting it right most of the time.
The more you train it the more accurate it gets.  There's no need to train
it on every message you receive, but you should train on a few spams and a
few hams on a regular basis.  You should also try to train it on about the
same number of spams as hams.

You can train it on lots of messages in one go by either using the
sb_filter.py script as explained in the "Command-line training" section,
or by giving messages to the web interface via the "Train" form on the
Home page. You can train on individual messages (which is tedious) or
using mbox files.


IMAP Filter
-----------

If you are running the IMAP filter with the '-t' switch, as described above,
then all you need to do to train is move examples of mail into the
appropriate folders, via your mail client (for example, move mail that was
not classified as spam into (one of) the folder(s) that you specified as
a spam training folder in the steps above.

Note that training, even without any classifying, using the IMAP filter,
means that your messages will be recreated (i.e. the old one is marked for
deletion and a new copy is made) on the server.  The messages will be
identical to the original, except that they will include an additional
header, so that SpamBayes can keep track of which messages have already
been processed.


Command-line training
---------------------

Given a pair of Unix mailbox format files (each message starts with a line
which begins with 'From '), one containing nothing but spam and the other
containing nothing but ham, you can train Spambayes using a command like

    sb_mboxtrain.py -g ~/tmp/newham -s ~/tmp/newspam

The above command is command-line-centric (eg. UNIX, or Windows command
prompt).  You can also use the web interface for training as detailed above.


Overview
========

[This section will tell you more about how and what SpamBayes is, but does
not contain any additional information about setting it up.]

There are eight main components to the SpamBayes system:

 o A database.  Loosely speaking, this is a collection of words and
   associated spam and ham probabilities.  The database says "If a message
   contains the word 'Viagra' then there's a 98% chance that it's spam, and
   a 2% chance that it's ham."  This database is created by training - you

CHANGELOG.txt
[Note that all dates are in ISO 8601 format, e.g. YYYY-MM-DD to ease sorting]

Release 3.1.0
=============

Mark Willson      2022-06-02  Merge changes from 1.1b3 ...
Mark Willson      2022-06-02  contrib/tte.py: catch UnicodeDecodeError exception
Mark Willson      2022-06-02  PKG-INFO: bump Metadata-Version and Version
Mark Willson      2022-06-02  scripts/sb_server.py: declare state global in _recreateState()
Mark Willson      2022-06-02  Added spambayes/asyncore.py and asynchat.py
Mark Willson      2022-06-02  spambayes/Dibbler.py: move import of asyncore, asynchat
Mark Willson      2022-06-02  spambayes/__init__.py: Update version information
Mark Willson      2022-06-02  spambayes/optimize.py: clobber __version__
Mark Willson      2022-06-02  spambayes/__init__.py: Update version information
Mark Willson      2022-06-02  spambayes/test/test_sb_imapfilter.py: asyncore now in spambayes
Mark Willson      2022-06-02  spambayes/test/test_sb_pop3dnd.py: asyncore now in spambayes
Mark Willson      2022-06-02  spambayes/test/test_sb_server.py: asyncore now in spambayes
Mark Willson      2022-06-02  spambayes/test/test_smtpproxy.py: asyncore now in spambayes
Mark Willson      2022-06-02  spambayes/tokenizer.py: catch ValueError when parsing url in class URLStripper.tokenize()
Mark Willson      2022-06-02  Deleted various files not in 1.1b3

Release 3.0.0
=============

Mark Willson      2020-03-13  Partial conversion to Python 3: scripts/sb_filter.py, scripts/sb_mboxtrain.py and supporting modules have been lightly tested.

Release 1.1a6
=============

Skip Montanaro    2008-12-06  Start requiring Python >= 2.4.
Skip Montanaro    2008-11-23  Route all pickle reads and writes through safepickle module.
Skip Montanaro    2008-11-23  Pick off a bunch of pylint nit (still tons to do).

Release 1.1a5
=============

Skip Montanaro	  2007-07-14  Add train and train_mime methods to XML-RPC plugin (from Marian Neagul).
Skip Montanaro    2007-07-15  Switch repository from CVS to Subversion.

Release 1.1a4
=============

Skip Montanaro    2006-09-09  Dump NetPBM decode support in favor of PIL
Skip Montanaro    2006-09-09  First crack at handling image sequences
Skip Montanaro    2007-05-11  IMAP patch for contrib/tte.py (Dave Abrahams)
Skip Montanaro    2007-05-11  Remove duplicate use of --cullext flag to contrib/tte.py
Skip Montanaro    2007-05-22  Note missing file name in error message - FileStorage.py
Skip Montanaro    2007-05-24  Set MinTTL to one day in dnscache.py
Skip Montanaro    2007-05-25  Catch broader exception in ImageStripper.py when image load fails (Sjoerd Mullender)
Skip Montanaro    2007-06-10  Add core_server.py & friends - plugin-based server

Release 1.1a3
=============

Skip Montanaro    2006-08-18  Update pycksum.py to try and identify more duplicates
Skip Montanaro	  2006-08-14  Add scale and charset options to ImageStripper
Skip Montanaro	  2006-08-13  Stitch spam images back together properly, add a couple more tokens
Skip Montanaro	  2006-08-10  Add support for PIL to ImageStripper.py
Skip Montanaro	  2006-08-09  Cache x-lookup_ip in a pickle instead of trying to use anydbm or zodb
Skip Montanaro	  2006-08-06  Add crude OCR capability to try and parse image-based spam using Ocrad & NetPBM
Skip Montanaro	  2006-08-06  Add x-short_runs option
Skip Montanaro	  2006-08-06  Add x-image_size option & corresponding token
Skip Montanaro	  2006-08-06  Add Matt Cowles' x-lookup_ip extension w/ slight modifications
Skip Montanaro	  2006-08-06  Add profiling using cProfile (if available) to sb_filter.py
Skip Montanaro	  2006-08-06  Delete -d and -p flags from spamcounts.py
Skip Montanaro	  2006-08-06  Refactor basic text tokenizing out of tokenize_body into a separate method, tokenize_text
Skip Montanaro	  2006-08-05  Explicitly close ZODB store in tte.py
Tony Meyer	  2006-06-22  Fix bug in regex preventing valid IPs
Toby Dickenson	  2006-06-12  Suppress spurious duplicate From_ lines in sb_bnfilter.py
Tony Meyer	  2006-06-10  Add simple parts of [ 824651 ] Multibyte message support
Tony Meyer	  2006-05-06  Enable -o command line option setting, and follow TestDriver directories in testtools/mksets.py
Skip Montanaro	  2006-04-23  Reduce sensitivity of spamcounts.py to classifier changes
Tony Meyer	  2006-04-22  Set zodb cache size to 10,000


Release 1.1a2
=============
Tony Meyer        2006-04-03  Add [ 1081787 ] Adding the version only to sb_filter.py
Tony Meyer        2006-04-03  Fix [ 1383801 ] trustedIPs wildcard to regex broken
Tony Meyer        2006-04-02  Fix [ 1387699 ] train_on_filter=True needs the db to be opened read/write
Tony Meyer        2006-04-02  Fix [ 1387709 ] If globals:dbm_type is non-default, then don't use whichdb.
Tony Meyer        2005-11-27  Install the conversion utility and offer to run it on Windows install.
Tony Meyer        2005-11-26  Add conversion utility to easily convert dbm to ZODB.
Tony Meyer        2005-11-26  Change database names to hammie.fs and messageinfo.fs
Tony Meyer        2005-11-26  Have ZODBClassifier honour read-only mode.
Tony Meyer        2005-11-26  Change default storage to ZODB.
Tony Meyer        2005-11-26  Remove backwards compatibility option to use True for dbm and False for pickle.
Tony Meyer        2005-11-26  Fix closing SSL connections in POP3 proxy.
Tony Meyer        2005-11-26  Abort failed ZODB transactions.
Tony Meyer        2005-11-15  Fix exception header adding.
Tony Meyer        2005-11-15  Tidy up the way that messages keep track of the messageinfo database.
Tony Meyer        2005-11-15  Add ZODB version of MessageInfo class.
Tony Meyer        2005-11-15  Let the notate_ options be turned off.
Tony Meyer        2005-11-15  Add simple handling of ZODB ConflictErrors.
Tony Meyer        2005-11-15  Pack ZODB database.
Tony Meyer        2005-11-15  Fix tokenizer so that the -o switch can be used with all options.
Tony Meyer        2005-11-15  Only run through expiry if messages might expire.

spambayes/Version.py
#! /usr/bin/env python
"""Simple version repository for SpamBayes core, and our main apps.

Also has the ability to load this version information from a remote location
(in that case, we actually load a "ConfigParser" version of the file to
avoid importing code we can't trust.)  This allows any app to check if there
is a later version available.

The makefile process for the website will execute this as a script, which
will generate the "ConfigParser" version for the web.
"""

import sys
import re

try:
    _
except NameError:
    _ = lambda arg: arg

# See bug 806238: urllib2 fails in Outlook new-version chk.
# A reason for why the spambayes.org URL fails is given in a comment there.
#LATEST_VERSION_HOME="http://www.spambayes.org/download/Version.cfg"
# The SF URL instead works for Tim and xenogeist.
LATEST_VERSION_HOME = "http://spambayes.sourceforge.net/download/Version.cfg"
DEFAULT_DOWNLOAD_PAGE = "http://spambayes.sourceforge.net/windows.html"

# This module is part of the spambayes project, which is Copyright 2002-2007
# The Python Software Foundation and is covered by the Python Software
# Foundation license.

versions = {
    # Note this means we can change the download page later, and old
    # versions will still go to the new page.
    # We may also like to have a "Release Notes Page" item later?
    "Download Page": DEFAULT_DOWNLOAD_PAGE,

    # Sub-dict for generating sub-sections in the cfg file that are compatible
    # with the update checking in older versions of SpamBayes.
    "Apps": {
        "Outlook" : {
            "Description":      "SpamBayes Outlook Addin",
        },
        "POP3 Proxy" : {
            "Description":      "SpamBayes POP3 Proxy",
        },
    },
}

def get_version(app = None,
                version_dict = None):
    """Get SBVersion object based on the version info in the supplied dict."""
    ver = SBVersion()  # get default version
    if version_dict is not None:
        dict = version_dict  # default to top level dictionary
        if app is not None:
            # attempt to get a sub-dict for the specific app
            try:
                dict = version_dict["Apps"][app]
            except KeyError:
                pass
        try:
            version = dict["Version"]
            # KLUDGE: Perform some bizarre magic to try to figure out if we
            # have an old-format float version instead of a new-format string
            # and massage it into a string format that will compare properly
            # in update checks.
            try:
                float(version)
                # Version converted successfully to a float, which means it
                # may be an old-format version number.  Old convention was to
                # use 1.01 to represent "1.0.1", so check to see if there is
                # more than one digit following the decimal.
                dot = version.find('.')
                ver_frac_part = version[dot+1:]
                if len(ver_frac_part) > 1:
                    # Use the first digit of the fractional part as the minor
                    # version and the rest as the patch version.
                    version = version[0:dot] + '.' + ver_frac_part[0] + '.' + ver_frac_part[1:]
            except ValueError:
                pass
            ver = SBVersion(version, version_dict["Date"])
        except KeyError:
            pass
    return ver

def get_download_page(app = None,
                      version_dict = None):
    if version_dict is None:
        version_dict = versions
    dict = version_dict  # default to top level dictionary
    if app is not None:
        # attempt to get a sub-dict for the specific app
        try:
            dict = version_dict["Apps"][app]
        except KeyError:
            pass
    try:
        return dict["Download Page"]
    except KeyError:
        # "Download Page" key not found so it may be an old-format dictionary.
        # Just use the default download page.
        return DEFAULT_DOWNLOAD_PAGE

def get_current_version():
    return SBVersion()

#============================================================================

# The SBVersion class is a modified version of the StrictVersion class from
# the "distutils" module.  It has been adapted to handle an "rc" pre-release
# designation for release candidates, and to store the version data in the
# format of the sys.version_info tuple.  A date string may also be provided
# that will be included in the long format of the version string.  The
# default version and date info is read from the metadata in the "spambayes"
# module __init__.py file.

class SBVersion:

    """Version numbering for SpamBayes releases.
    A version number consists of two or three dot-separated numeric
    components, with an optional "pre-release" tag on the end.  The
    pre-release tag consists of the designations 'a' (for alpha),
    'b' (for beta), or 'rc' (for release candidate) followed by a number.
    If the numeric components of two version numbers are equal, then one
    with a pre-release tag will always be deemed earlier (lesser) than
    one without.

    The following are valid version numbers (shown in the order that
    would be obtained by sorting according to the supplied cmp function):

        0.4       0.4.0  (these two are equivalent)
        0.4.1
        0.5a1
        0.5b3
        0.5
        0.9.6
        1.0
        1.0.4a3
        1.0.4b1
        1.0.4rc2
        1.0.4

    The following are examples of invalid version numbers:

        1
        2.7.2.2
        1.3.a4
        1.3pl1
        1.3c4

    A date may also be associated with the version, typically to track the
    date when the release was made public.  The date is specified as a string,
    and is only used in formatting the long version of the version string.
    """

    def __init__(self, vstring=None, date=None):
        import spambayes
        if vstring:
            self.parse(vstring)
        else:
            self.parse(spambayes.__version__)
        if date:
            self.date = date
        else:
            self.date = spambayes.__date__

    def __repr__ (self):
        return "%s('%s', '%s')" % (self.__class__.__name__, str(self), self.date)

    version_re = re.compile(r'^(\d+) \. (\d+) (\. (\d+))? (([ab]|rc)(\d+))?\+?$',
                            re.VERBOSE)

    def parse(self, vstring):
        match = self.version_re.match(vstring)
        if not match:
            raise ValueError("invalid version number '%s'" % vstring)

        (major, minor, patch, prerelease, prerelease_num) = \
            match.group(1, 2, 4, 6, 7)

        if not patch:
            patch = "0"
            
        if not prerelease:
            releaselevel = "final"
            serial = 0
        else:
            serial = int(prerelease_num)
            if prerelease == "a":
                releaselevel = "alpha"
            elif prerelease == "b":
                releaselevel = "beta"
            elif prerelease == "rc":
                releaselevel = "candidate"
        self.version_info = tuple(list(map(int, [major, minor, patch])) + \
                                  [releaselevel, serial])

    def __str__(self):
        if self.version_info[2] == 0:
            vstring = '.'.join(map(str, self.version_info[0:2]))
        else:
            vstring = '.'.join(map(str, self.version_info[0:3]))

        releaselevel = self.version_info[3][0]
        if releaselevel != 'f':
            if releaselevel == 'a':
                prerelease = "a"
            elif releaselevel == 'b':
                prerelease = "b"
            elif releaselevel == 'c':
                prerelease = "rc"
            vstring = vstring + prerelease + str(self.version_info[4])

        return vstring

    def __cmp__(self, other):
        if isinstance(other, str):
            other = SBVersion(other)

        return cmp(self.version_info, other.version_info)

    def get_long_version(self, app_name = None):
        if app_name is None:
            app_name = "SpamBayes"
        return _("%s Version %s (%s)") % (app_name, str(self), self.date)

#============================================================================

# Utilities to check the "latest" version of an app.
# Assumes that a 'config' version of this file exists at the given URL
# No exceptions are caught
try:
    import configparser
    class MySafeConfigParser(configparser.SafeConfigParser):
        def optionxform(self, optionstr):
            return optionstr # no lower!
except AttributeError: # No SafeConfigParser!
    MySafeConfigParser = None

def fetch_latest_dict(url=LATEST_VERSION_HOME):
    if MySafeConfigParser is None:
        raise RuntimeError("Sorry, but only Python 2.3+ can trust remote config files")

    import urllib.request, urllib.error, urllib.parse
    from spambayes.Options import options
    server = options["globals", "proxy_server"]
    if server != "":
        if ':' in server:
            server, port = server.split(':', 1)
            port = int(port)
        else:
            port = 8080
        if options["globals", "proxy_username"]:
            user_pass_string = "%s:%s" % \
                               (options["globals", "proxy_username"],
                                options["globals", "proxy_password"])
        else:
            user_pass_string = ""
        proxy_support = urllib.request.ProxyHandler({"http" :
                                              "http://%s@%s:%d" % \
                                              (user_pass_string, server,
                                               port)})
        opener = urllib.request.build_opener(proxy_support, urllib.request.HTTPHandler)
        urllib.request.install_opener(opener)
    stream = urllib.request.urlopen(url)
    cfg = MySafeConfigParser()
    cfg.readfp(stream)
    ret_dict = {}
    apps_dict = ret_dict["Apps"] = {}
    for sect in cfg.sections():
        if sect == "SpamBayes":
            target_dict = ret_dict
        else:
            target_dict = apps_dict.setdefault(sect, {})
        for opt in cfg.options(sect):
            val = cfg.get(sect, opt)
            target_dict[opt] = val
    return ret_dict

# Utilities for generating a 'config' version of this file.
# The output of this should exist at the URL above.
compat_apps = {
    "Outlook" : {
        "Description":      "SpamBayes Outlook Addin",
    },
    "POP3 Proxy" : {
        "Description":      "SpamBayes POP3 Proxy",
    },
}
shared_cfg_opts = {
    # Note this means we can change the download page later, and old
    # versions will still go to the new page.
    # We may also like to have a "Release Notes Page" item later?
    "Download Page": "http://spambayes.sourceforge.net/windows.html",
}
def _write_cfg_opts(stream, this_dict):
    for name, val in list(this_dict.items()):
        if type(val)==type(''):
            val_str = repr(val)[1:-1]
        elif type(val)==type(0.0):
            val_str = str(val)
        elif type(val)==type({}):
            val_str = None # sub-dict
        else:
            print(("Skipping unknown value type: %r" % val))
            val_str = None
        if val_str is not None:
            stream.write("%s:%s\n" % (name, val_str))
def _make_compatible_cfg_section(stream, key, ver, this_dict):
    stream.write("[%s]\n" % key)
    # We need to create a float representation of the current version that
    # sort correctly in older versions that used a float version number.
    ver_num = float(ver.version_info[0])
    ver_num += float(ver.version_info[1] * 0.1)
    ver_num += float(ver.version_info[2] * 0.01)
    releaselevel = ver.version_info[3][0]
    if releaselevel == 'a':
        prerelease_offset = 0.001 - (float(ver.version_info[4]) * 0.00001)
    elif releaselevel == 'b':
        prerelease_offset = 0.0005 - (float(ver.version_info[4]) * 0.00001)
    elif releaselevel == 'c':
        prerelease_offset = 0.0001 - (float(ver.version_info[4]) * 0.00001)
    else:
        prerelease_offset = 0.0
    ver_num -= prerelease_offset
    stream.write("Version:%s\n" % str(ver_num))
    stream.write("BinaryVersion:%s\n" % str(ver_num))
    stream.write("Date:%s\n" % ver.date)
    _write_cfg_opts(stream, this_dict)
    desc_str = "%%(Description)s Version %s (%%(Date)s)" % str(ver)
    stream.write("Full Description:%s\n" % desc_str)
    stream.write("Full Description Binary:%s\n" % desc_str)
    _write_cfg_opts(stream, versions)
    stream.write("\n")
def _make_cfg_section(stream, ver):
    stream.write("[SpamBayes]\n")
    stream.write("Version:%s\n" % str(ver))
    stream.write("Date:%s\n" % ver.date)
    _write_cfg_opts(stream, versions)
    stream.write("\n")

def make_cfg(stream):
    stream.write("# This file is generated from spambayes/Version.py" \
                 " - do not edit\n")
    ver = get_current_version()
    _make_cfg_section(stream, ver)
    for appname in compat_apps:
        _make_compatible_cfg_section(stream, appname, ver, versions["Apps"][appname])

def main(args):
    if '-g' in args:
        make_cfg(sys.stdout)
        sys.exit(0)
        
    v_this = get_current_version()
    print(("Current version:", v_this.get_long_version()))

    print()
    print("Fetching the lastest version information...")
    try:
        latest_dict = fetch_latest_dict()
    except:
        print("FAILED to fetch the latest version")
        import traceback
        traceback.print_exc()
        sys.exit(1)

    v_latest = get_version(version_dict=latest_dict)
    print()
    print(("Latest version:", v_latest.get_long_version()))

if __name__ == '__main__':
    main(sys.argv)
